File: NovelSpeaker/Speaker.m
//
//  Speaker.m
//  NBackSpeaker
//
//  Created by 飯村 卓司 on 2014/02/16.
//  Copyright (c) 2014年 IIMURA Takuji. All rights reserved.
//

#import "Speaker.h"

@implementation Speaker

@synthesize speakRangeChangeDelegate;

- (id) init {
    self = [super init];
    if (self == nil) {
        return nil;
    }
    
    m_Synthesizer = [AVSpeechSynthesizer new];
    m_Voice = [AVSpeechSynthesisVoice voiceWithLanguage:@"ja-JP"];
    m_Rate = 0.7f;
    m_Pitch = 1.0f;
    m_Interval = 0.0;
    m_Volume = 1.0f;
    m_AudioEngine = nil;
    m_AudioPlayerNode = nil;
    m_AudioFormat_After = nil;
    m_AudioFormat_Before = nil;
    m_AudioConverter = nil;
    m_CurrentStatus = STSpeakingStatusNone;

    m_Synthesizer.delegate = self;

    // AVAudioSessionInterruptionNotification のイベントハンドラを登録します
    [self setAVAudioSessionInterruptionNotificationHandler];
    
    return self;
}
    
- (void)dealloc
{
    NSNotificationCenter* center = [NSNotificationCenter defaultCenter];
    [center removeObserver:self];
    if (m_Synthesizer != nil) {
        m_Synthesizer.delegate = nil;
    }
}

- (AVSpeechUtterance*)CreateUtterance:(NSString*)text {
    AVSpeechUtterance *utterance = [AVSpeechUtterance speechUtteranceWithString:text];
    if (m_Voice == nil) {
        m_Voice = [AVSpeechSynthesisVoice voiceWithLanguage:@"ja-JP"];
    }
    utterance.voice = m_Voice;
    utterance.rate = m_Rate;
    if (m_Volume > 0.0 && m_Volume < 1.0) {
        utterance.volume = m_Volume;
    }
    utterance.pitchMultiplier = m_Pitch;
    utterance.postUtteranceDelay = m_Interval;
    return utterance;
}

- (void)SetupAudioEngine:(AVAudioFormat*)format volume:(float)volume {
    // format を変更するための AVAudioConverter が用意されているならそのまま返します
    if (m_AudioFormat_After != nil
        && m_AudioFormat_Before != nil
        && m_AudioFormat_Before.commonFormat == format.commonFormat
        && m_AudioFormat_After.channelCount == format.channelCount
        && m_AudioFormat_After.interleaved == format.interleaved
        && fabs(m_AudioFormat_After.sampleRate - format.sampleRate) < DBL_EPSILON) {
        return;
    }
    // format を変更する必要がなくて、AudioPlayerNode が用意されているならそのまま返します
    if ((format.commonFormat == AVAudioPCMFormatFloat32
        || format.commonFormat == AVAudioPCMFormatFloat64)
        && m_AudioEngine != nil
        && m_AudioPlayerNode != nil) {
        return;
    }
    NSLog(@"create AudioEngine:\n\tchannelCount: %u -> %u\n\tinterleaved: %@ -> %@\n\tsampleRate: %f -> %f, volume: %f"
          , m_AudioFormat_After.channelCount, format.channelCount
          , m_AudioFormat_After.interleaved ? @"True" : @"False", format.interleaved ? @"True" : @"False"
          , m_AudioFormat_After.sampleRate, format.sampleRate
          , m_Volume);
    if (m_AudioEngine != nil) {
        [m_AudioEngine stop];
    }
    m_AudioEngine = [AVAudioEngine new];
    m_AudioPlayerNode = [AVAudioPlayerNode new];
    [m_AudioEngine attachNode:m_AudioPlayerNode];
    AVAudioMixerNode* mixerNode = [m_AudioEngine mainMixerNode];
    // なんでだかわからないけれど、AVAudioEngine で playerNode と mixerNode を繋ぐ時、
    // format として AVAudioPCMFormatInt16 を指定するとそこで落ちるため
    // それらであれば AVAudioPCMFormatFloat32 に変換するように converter を用意します。
    AVAudioFormat* toFormat = format;
    if (format.commonFormat == AVAudioPCMFormatInt16
        || format.commonFormat == AVAudioPCMFormatInt32) {
        AVAudioFormat* fromFormat = format;
        toFormat = [[AVAudioFormat alloc] initWithCommonFormat:AVAudioPCMFormatFloat32 sampleRate:fromFormat.sampleRate channels:fromFormat.channelCount interleaved:fromFormat.interleaved];
        AVAudioConverter* converter = [[AVAudioConverter alloc] initFromFormat:fromFormat toFormat:toFormat];
        m_AudioFormat_After = toFormat;
        m_AudioFormat_Before = fromFormat;
        m_AudioConverter = converter;
    }else{
        m_AudioConverter = nil;
    }
    [m_AudioEngine connect:m_AudioPlayerNode to:mixerNode format:toFormat];
    [m_AudioPlayerNode setVolume:volume];
    [m_AudioEngine startAndReturnError:nil];
    [m_AudioPlayerNode play];
}

- (BOOL) SpeechToBuffer:(NSString*)text {
    AVSpeechUtterance* utterance = [self CreateUtterance:text];
    
    if (@available(iOS 13.0, *)) {
        [m_Synthesizer writeUtterance:utterance toBufferCallback:^(AVAudioBuffer * _Nonnull buffer) {
            if (![buffer isKindOfClass:[AVAudioPCMBuffer class]]) {
                return;
            }
            AVAudioPCMBuffer* _Nonnull pcmBuffer = (AVAudioPCMBuffer* _Nonnull)buffer;
            if (pcmBuffer.frameLength <= 0) {
                return;
            }
            AVAudioFormat* format = [pcmBuffer format];
            [self SetupAudioEngine:format volume:m_Volume];
            @autoreleasepool {
                AVAudioPCMBuffer* buffer = pcmBuffer;
                if (m_AudioConverter != nil
                    && format.commonFormat == m_AudioFormat_Before.commonFormat) {
                    AVAudioPCMBuffer* newBuffer = [[AVAudioPCMBuffer alloc] initWithPCMFormat:m_AudioFormat_After frameCapacity:pcmBuffer.frameLength];
                    NSError* err;
                    [m_AudioConverter convertToBuffer:newBuffer fromBuffer:pcmBuffer error:&err];
                    buffer = newBuffer;
                }
                [m_AudioPlayerNode scheduleBuffer:buffer completionHandler:nil];
            }
        }];
    } else {
        return [self Speech:text];
    }
    return TRUE;
}

- (BOOL) Speech: (NSString*) text
{
    if (@available(iOS 13.0, *)) {
        if (m_Volume > 1.0) {
            return [self SpeechToBuffer:text];
        }
    }
    // memo: AVSpeechSynthesizer:speakUtterance は再生queueに追加される形式のようなので、再生中でも追加してかまわないっぽいです
    AVSpeechUtterance* utterance = [self CreateUtterance:text];

    //NSLog(@"rate: %f, pitch: %f, post delay: %f text: %@", m_Rate, m_Pitch, m_Interval, text);
    [m_Synthesizer speakUtterance:utterance];
    
    return TRUE;
}

- (void) SetVoice: (NSString*) language
{
    m_Voice = [AVSpeechSynthesisVoice voiceWithLanguage:language];
}

- (BOOL)SetVoiceWithIdentifier:(NSString*) voiceID voiceLocale:(NSString*)voiceLocale
{
    if (NSFoundationVersionNumber < NSFoundationVersionNumber_iOS_9_0) {
        return false;
    }
    if (voiceID == nil) {
        return false;
    }
    AVSpeechSynthesisVoice* voice = [AVSpeechSynthesisVoice voiceWithIdentifier:voiceID];
    if (voice == nil) {
        NSLog(@"can not set voiceIdentifier: %@. try fallback to locale: %@", voiceID, voiceLocale);
        voice = [AVSpeechSynthesisVoice voiceWithLanguage:voiceLocale];
        if (voice == nil) {
            NSLog(@"can not set voiceIdentifier on locale: %@.", voiceLocale);
        }
        return false;
    }
    m_Voice = voice;
    return true;
}

- (void) SetRate: (float) rate
{
    if(rate < AVSpeechUtteranceMinimumSpeechRate)
    {
        rate = AVSpeechUtteranceMinimumSpeechRate;
    }else if(rate > AVSpeechUtteranceMaximumSpeechRate)
    {
        rate = AVSpeechUtteranceMaximumSpeechRate;
    }
    m_Rate = rate;
}

- (void) SetPitch: (float) pitch
{
    m_Pitch = pitch;
}

- (void) SetDelay: (NSTimeInterval) interval
{
    m_Interval = interval;
}

/// 音声を読み上げる時の音量を指定します。
/// 1.0以上の値を指定すると、iOS13以降又はwatchOS 6以降の機能を利用するため、
/// それ以前のOSで1.0以上の値を指定しても無視され、1.0として扱われます。
/// また、負の値を指定すると無視されます。
- (void) SetVolume: (float) volume{
    m_Volume = volume;
}


- (STSpeakingStatus) GetStatus
{
    return m_CurrentStatus;
}

- (BOOL) StopSpeech
{
    //if ([m_Synthesizer isSpeaking]) {
    //NSLog(@"#### AVSpeech Stop Immediate####");
        [m_Synthesizer stopSpeakingAtBoundary:AVSpeechBoundaryImmediate];
    //}

    /*
    // なにやら読み上げが失敗するようになることがあるようなので、Stopのタイミングで AVSpeechSynthesizer object を作り直すようにします
    m_Synthesizer.delegate = nil; // 先に delegate は消しておきます
    m_Synthesizer = nil; // 走れ！走れ！AVSpeechSynthesizerのデストラクタよ走れ！
    m_Synthesizer = [AVSpeechSynthesizer new];
    m_Synthesizer.delegate = self;
     */
    
    return true;
}

- (BOOL) PauseSpeech
{
    return [m_Synthesizer pauseSpeakingAtBoundary:AVSpeechBoundaryImmediate];
}

- (BOOL) ResumeSpeech
{
    return [m_Synthesizer continueSpeaking];
}

- (void)changeStatus:(STSpeakingStatus)status
{
    m_CurrentStatus = status;
    switch(status)
    {
        case STSpeakingStatusNone:
            //NSLog(@"speak status -> None");
            break;
        case STSpeakingStatusPause:
            NSLog(@"speak status -> Pause");
            break;
        case STSpeakingStatusSpeak:
            //NSLog(@"speak status -> Speak");
            break;
        case STSpeakingStatusStop:
            //NSLog(@"speak status -> Stop");
            if (self.speakRangeChangeDelegate != nil) {
                [self.speakRangeChangeDelegate finishSpeak];
            }
            break;
        default:
            NSLog(@"speak status -> Unknown");
            break;
    }
}

#pragma mark - AVSpeechSynthesizerDelegate

- (void)speechSynthesizer:(AVSpeechSynthesizer *)synthesizer didStartSpeechUtterance:(AVSpeechUtterance *)utterance
{
	[self changeStatus:STSpeakingStatusSpeak];
}
- (void)speechSynthesizer:(AVSpeechSynthesizer *)synthesizer didFinishSpeechUtterance:(AVSpeechUtterance *)utterance{
	[self changeStatus:STSpeakingStatusStop];
}
- (void)speechSynthesizer:(AVSpeechSynthesizer *)synthesizer didPauseSpeechUtterance:(AVSpeechUtterance *)utterance{
	[self changeStatus:STSpeakingStatusPause];
}
- (void)speechSynthesizer:(AVSpeechSynthesizer *)synthesizer didContinueSpeechUtterance:(AVSpeechUtterance *)utterance{
	[self changeStatus:STSpeakingStatusSpeak];
}
- (void)speechSynthesizer:(AVSpeechSynthesizer *)synthesizer didCancelSpeechUtterance:(AVSpeechUtterance *)utterance{
	[self changeStatus:STSpeakingStatusStop];
}

// watchOS でこれを見えるようにした状態で AVSpeechSynthesizer に delegate を登録すると、読み上げ時にメモリを 30MBytes から 40MBytes 位持っていかれて、しかも開放してくれなくなるので封印します。(´・ω・`)
#if TARGET_OS_IOS
- (void)speechSynthesizer:(AVSpeechSynthesizer *)synthesizer willSpeakRangeOfSpeechString:(NSRange)characterRange utterance:(AVSpeechUtterance *)utterance{
    if (self.speakRangeChangeDelegate == nil) {
        return;
    }
    [self.speakRangeChangeDelegate willSpeakRange:characterRange speakText:utterance.speechString];
}
#endif

- (void)setAVAudioSessionInterruptionNotificationHandler
{
    NSNotificationCenter* center = [NSNotificationCenter defaultCenter];
    [center addObserver:self selector:@selector(sessionDidInterrupt:) name:AVAudioSessionInterruptionNotification object:nil];
}

- (void)sessionDidInterrupt:(NSNotification*)notification
{
    if (notification == nil) {
        return;
    }
    
    switch ([notification.userInfo[AVAudioSessionInterruptionTypeKey] intValue]) {
        case AVAudioSessionInterruptionTypeBegan:
            NSLog(@"interruption begin.");
            [self PauseSpeech];
            break;
        case AVAudioSessionInterruptionTypeEnded:
            NSLog(@"interuption end.");
            [self ResumeSpeech];
            break;
        default:
            break;
    }
}

@end

File: NovelSpeaker/Speaker.swift
//
//  SpeakerSwift.swift
//  NovelSpeaker
//
//  Created by 飯村卓司 on 2020/01/10.
//  Copyright © 2020 IIMURA Takuji. All rights reserved.
//

import Foundation
import AVFoundation

protocol SpeakRangeDelegate {
    func willSpeakRange(range:NSRange)
    func finishSpeak(isCancel:Bool, speechString:String)
}

class Speaker {
    var speaker_Original:Speaker_Original? = nil
    var speaker_WithoutWillSpeakRange:Speaker_WithoutWillSpeakRange? = nil
    
    init() {
        AssignSpeaker()
    }
    
    func AssignSpeaker() {
        if NovelSpeakerUtility.GetIsDisableWillSpeakRange() {
            if speaker_Original == nil && speaker_WithoutWillSpeakRange != nil {
                return
            }
            speaker_Original = nil
            speaker_WithoutWillSpeakRange = Speaker_WithoutWillSpeakRange()
        }else{
            if speaker_Original != nil && speaker_WithoutWillSpeakRange == nil {
                return
            }
            speaker_Original = Speaker_Original()
            speaker_WithoutWillSpeakRange = nil
        }
    }
    
    func Speech(text:String) {
        speaker_Original?.Speech(text: text)
        speaker_WithoutWillSpeakRange?.Speech(text: text)
    }
    
    func Stop() {
        speaker_Original?.Stop()
        speaker_WithoutWillSpeakRange?.Stop()
    }
    
    func Pause() {
        speaker_Original?.Pause()
        speaker_WithoutWillSpeakRange?.Pause()
    }
    
    func Resume() {
        speaker_Original?.Resume()
        speaker_WithoutWillSpeakRange?.Resume()
    }
    
    var voice:AVSpeechSynthesisVoice {
        get {
            if let speaker = speaker_Original {
                return speaker.voice
            }
            if let speaker = speaker_WithoutWillSpeakRange {
                return speaker.voice
            }
            return AVSpeechSynthesisVoice()
        }
        set(value) {
            speaker_Original?.voice = value
            speaker_WithoutWillSpeakRange?.voice = value
        }
    }
    func SetVoiceWith(identifier:String, language:String) {
        speaker_Original?.SetVoiceWith(identifier: identifier, language: language)
        speaker_WithoutWillSpeakRange?.SetVoiceWith(identifier: identifier, language: language)
    }
    func SetVoiceWith(language:String) {
        speaker_Original?.SetVoiceWith(language: language)
        speaker_WithoutWillSpeakRange?.SetVoiceWith(language: language)
    }
    var pitch:Float {
        get {
            if let speaker = speaker_Original {
                return speaker.pitch
            }
            if let speaker = speaker_WithoutWillSpeakRange {
                return speaker.pitch
            }
            return 1.0
        }
        set(value) {
            speaker_Original?.pitch = value
            speaker_WithoutWillSpeakRange?.pitch = value
        }
    }
    var rate:Float {
        get {
            if let speaker = speaker_Original {
                return speaker.rate
            }
            if let speaker = speaker_WithoutWillSpeakRange {
                return speaker.rate
            }
            return 1.0
        }
        set(value) {
            speaker_Original?.rate = value
            speaker_WithoutWillSpeakRange?.rate = value
        }
    }
    var volume:Float {
        get {
            if let speaker = speaker_Original {
                return speaker.volume
            }
            if let speaker = speaker_WithoutWillSpeakRange {
                return speaker.volume
            }
            return 1.0
        }
        set(value) {
            speaker_Original?.volume = value
            speaker_WithoutWillSpeakRange?.volume = value
        }
    }
    var delay:TimeInterval {
        get {
            if let speaker = speaker_Original {
                return speaker.delay
            }
            if let speaker = speaker_WithoutWillSpeakRange {
                return speaker.delay
            }
            return 1.0
        }
        set(value) {
            speaker_Original?.delay = value
            speaker_WithoutWillSpeakRange?.delay = value
        }
    }
    var delegate:SpeakRangeDelegate? {
        get {
            if let speaker = speaker_Original {
                return speaker.delegate
            }
            if let speaker = speaker_WithoutWillSpeakRange {
                return speaker.delegate
            }
            return nil
        }
        set(value) {
            speaker_Original?.delegate = value
            speaker_WithoutWillSpeakRange?.delegate = value
        }
    }

    func isSpeaking() -> Bool {
        if let speaker = speaker_Original {
            return speaker.isSpeaking()
        }
        if let speaker = speaker_WithoutWillSpeakRange {
            return speaker.isSpeaking()
        }
        return false
    }
    
    func reloadSynthesizer() {
        speaker_Original?.reloadSynthesizer()
        speaker_WithoutWillSpeakRange?.reloadSynthesizer()
    }
    
    func ChangeSpeakerWillSpeakRangeType() {
        let prevSpeakerOriginal = speaker_Original
        let prevSpeakerWithoutWillSpeakRange = speaker_WithoutWillSpeakRange
        AssignSpeaker()
        if let speaker = speaker_Original, let prev = prevSpeakerWithoutWillSpeakRange {
            speaker.delegate = prev.delegate
            speaker.voice = prev.voice
            speaker.delay = prev.delay
            speaker.volume = prev.volume
            speaker.rate = prev.rate
            speaker.pitch = prev.pitch
        }
        if let speaker = speaker_WithoutWillSpeakRange, let prev = prevSpeakerOriginal {
            speaker.delegate = prev.delegate
            speaker.voice = prev.voice
            speaker.delay = prev.delay
            speaker.volume = prev.volume
            speaker.rate = prev.rate
            speaker.pitch = prev.pitch
        }
    }
    
    func isPaused() -> Bool {
        if let speaker = speaker_Original {
            return speaker.isPaused()
        }
        if let speaker = speaker_WithoutWillSpeakRange {
            return speaker.isPaused()
        }
        return false
    }

    var isSpeechKicked:Bool {
        get {
            if let speaker = speaker_Original {
                return speaker.isSpeechKicked
            }
            if let speaker = speaker_WithoutWillSpeakRange {
                return speaker.isSpeechKicked
            }
            return false
        }
    }
}

class Speaker_Original: NSObject, AVSpeechSynthesizerDelegate {
    var synthesizer = AVSpeechSynthesizer()
    var m_Voice:AVSpeechSynthesisVoice = AVSpeechSynthesisVoice(language: "ja-JP") ?? AVSpeechSynthesisVoice()
    var m_Pitch:Float = 1.0
    var m_Rate:Float = AVSpeechUtteranceDefaultSpeechRate
    var m_Volume:Float = 1.0
    var m_Delay:TimeInterval = 0.0
    var m_Delegate:SpeakRangeDelegate? = nil
    var isSpeechKicked:Bool = false
    
    override init() {
        super.init()
        synthesizer.delegate = self
        SetNotificationHandler()
    }
    
    deinit {
        RemoveNotificationHandler()
        synthesizer.delegate = nil
    }

    func Speech(text:String) {
        if NiftyUtility.isTesting() {
            return
        }
        isSpeechKicked = true
        let utt = AVSpeechUtterance(string: text)
        utt.voice = m_Voice
        utt.pitchMultiplier = m_Pitch
        utt.rate = m_Rate
        utt.postUtteranceDelay = m_Delay
        utt.volume = max(0.0, min(1.0, m_Volume))
        synthesizer.speak(utt)
    }
    
    func Stop() {
        synthesizer.stopSpeaking(at: .immediate)
    }
    
    func Pause() {
        synthesizer.pauseSpeaking(at: .immediate)
    }
    
    func Resume() {
        synthesizer.continueSpeaking()
    }
    
    var voice:AVSpeechSynthesisVoice {
        get { return m_Voice }
        set(value) { m_Voice = value }
    }
    func SetVoiceWith(identifier:String, language:String) {
        if let voice = AVSpeechSynthesisVoice(identifier: identifier) {
            m_Voice = voice
            return
        }
        if let voice = AVSpeechSynthesisVoice(language: language) {
            m_Voice = voice
            return
        }
    }
    func SetVoiceWith(language:String) {
        if let voice = AVSpeechSynthesisVoice(language: language) {
            m_Voice = voice
        }
    }
    var pitch:Float {
        get { return m_Pitch }
        set(value) { m_Pitch = value }
    }
    var rate:Float {
        get { return m_Rate }
        set(value) {
            if value > AVSpeechUtteranceMaximumSpeechRate {
                m_Rate = AVSpeechUtteranceMaximumSpeechRate
                return
            }
            if value < AVSpeechUtteranceMinimumSpeechRate {
                m_Rate = AVSpeechUtteranceMinimumSpeechRate
                return
            }
            m_Rate = value
        }
    }
    var volume:Float {
        get { return m_Volume }
        set(value) {
            m_Volume = max(0.0, min(1.0, value))
        }
    }
    var delay:TimeInterval {
        get { return m_Delay }
        set(value) { m_Delay = value }
    }
    var delegate:SpeakRangeDelegate? {
        get { return m_Delegate }
        set(value) { m_Delegate = value }
    }
    
    func SetNotificationHandler() {
        let center = NotificationCenter.default
        center.addObserver(self, selector: #selector(sessionDidInterrupt(notification:)), name: AVAudioSession.interruptionNotification, object: nil)
    }
    func RemoveNotificationHandler() {
        let center = NotificationCenter.default
        center.removeObserver(self)
    }
    
    @objc func sessionDidInterrupt(notification:Notification) {
        guard let interruptType = notification.userInfo?[AVAudioSessionInterruptionTypeKey] as? NSNumber, let type = AVAudioSession.InterruptionType(rawValue: interruptType.uintValue) else { return }
        switch type {
        case AVAudioSession.InterruptionType.began:
            Pause()
        case AVAudioSession.InterruptionType.ended:
            Resume()
        default:
            break
        }
    }
    
    @objc func speechSynthesizer(_ synthesizer: AVSpeechSynthesizer, didFinish utterance: AVSpeechUtterance) {
        delegate?.finishSpeak(isCancel: false, speechString: utterance.speechString)
    }
    @objc func speechSynthesizer(_ synthesizer: AVSpeechSynthesizer, didCancel utterance: AVSpeechUtterance) {
        delegate?.finishSpeak(isCancel: true, speechString: utterance.speechString)
    }

    #if !os(watchOS)
    @objc func speechSynthesizer(_ synthesizer: AVSpeechSynthesizer, willSpeakRangeOfSpeechString characterRange: NSRange, utterance: AVSpeechUtterance) {
        if let delegate = self.m_Delegate {
            delegate.willSpeakRange(range: characterRange)
        }
    }
    #endif
    
    func isSpeaking() -> Bool {
        return self.synthesizer.isSpeaking
    }
    
    func reloadSynthesizer() {
        synthesizer = AVSpeechSynthesizer()
        synthesizer.delegate = self
    }
    
    func isPaused() -> Bool {
        return self.synthesizer.isPaused
    }
}
class Speaker_WithoutWillSpeakRange: NSObject, AVSpeechSynthesizerDelegate {
    var synthesizer = AVSpeechSynthesizer()
    var m_Voice:AVSpeechSynthesisVoice = AVSpeechSynthesisVoice(language: "ja-JP") ?? AVSpeechSynthesisVoice()
    var m_Pitch:Float = 1.0
    var m_Rate:Float = AVSpeechUtteranceDefaultSpeechRate
    var m_Volume:Float = 1.0
    var m_Delay:TimeInterval = 0.0
    var m_Delegate:SpeakRangeDelegate? = nil
    var isSpeechKicked:Bool = false
    
    override init() {
        super.init()
        synthesizer.delegate = self
        SetNotificationHandler()
    }
    
    deinit {
        RemoveNotificationHandler()
        synthesizer.delegate = nil
    }
    
    func Speech(text:String) {
        if NiftyUtility.isTesting() {
            return
        }
        isSpeechKicked = true
        let utt = AVSpeechUtterance(string: text)
        utt.voice = m_Voice
        utt.pitchMultiplier = m_Pitch
        utt.rate = m_Rate
        utt.postUtteranceDelay = m_Delay
        utt.volume = max(0.0, min(1.0, m_Volume))
        synthesizer.speak(utt)
    }
    
    func Stop() {
        synthesizer.stopSpeaking(at: .immediate)
    }
    
    func Pause() {
        synthesizer.pauseSpeaking(at: .immediate)
    }
    
    func Resume() {
        synthesizer.continueSpeaking()
    }
    
    var voice:AVSpeechSynthesisVoice {
        get { return m_Voice }
        set(value) { m_Voice = value }
    }
    func SetVoiceWith(identifier:String, language:String) {
        if let voice = AVSpeechSynthesisVoice(identifier: identifier) {
            m_Voice = voice
            return
        }
        if let voice = AVSpeechSynthesisVoice(language: language) {
            m_Voice = voice
            return
        }
    }
    func SetVoiceWith(language:String) {
        if let voice = AVSpeechSynthesisVoice(language: language) {
            m_Voice = voice
        }
    }
    var pitch:Float {
        get { return m_Pitch }
        set(value) { m_Pitch = value }
    }
    var rate:Float {
        get { return m_Rate }
        set(value) {
            if value > AVSpeechUtteranceMaximumSpeechRate {
                m_Rate = AVSpeechUtteranceMaximumSpeechRate
                return
            }
            if value < AVSpeechUtteranceMinimumSpeechRate {
                m_Rate = AVSpeechUtteranceMinimumSpeechRate
                return
            }
            m_Rate = value
        }
    }
    var volume:Float {
        get { return m_Volume }
        set(value) {
            m_Volume = max(0.0, min(1.0, value))
        }
    }
    var delay:TimeInterval {
        get { return m_Delay }
        set(value) { m_Delay = value }
    }
    var delegate:SpeakRangeDelegate? {
        get { return m_Delegate }
        set(value) { m_Delegate = value }
    }
    
    func SetNotificationHandler() {
        let center = NotificationCenter.default
        center.addObserver(self, selector: #selector(sessionDidInterrupt(notification:)), name: AVAudioSession.interruptionNotification, object: nil)
    }
    func RemoveNotificationHandler() {
        let center = NotificationCenter.default
        center.removeObserver(self)
    }
    
    @objc func sessionDidInterrupt(notification:Notification) {
        guard let interruptType = notification.userInfo?[AVAudioSessionInterruptionTypeKey] as? NSNumber, let type = AVAudioSession.InterruptionType(rawValue: interruptType.uintValue) else { return }
        switch type {
        case AVAudioSession.InterruptionType.began:
            Pause()
        case AVAudioSession.InterruptionType.ended:
            Resume()
        default:
            break
        }
    }
    
    @objc func speechSynthesizer(_ synthesizer: AVSpeechSynthesizer, didFinish utterance: AVSpeechUtterance) {
        delegate?.finishSpeak(isCancel: false, speechString: utterance.speechString)
    }
    @objc func speechSynthesizer(_ synthesizer: AVSpeechSynthesizer, didCancel utterance: AVSpeechUtterance) {
        delegate?.finishSpeak(isCancel: true, speechString: utterance.speechString)
    }

    /*
    #if !os(watchOS)
    @objc func speechSynthesizer(_ synthesizer: AVSpeechSynthesizer, willSpeakRangeOfSpeechString characterRange: NSRange, utterance: AVSpeechUtterance) {
        if let delegate = self.m_Delegate {
            delegate.willSpeakRange(range: characterRange)
        }
    }
    #endif
     */
    
    func isSpeaking() -> Bool {
        return self.synthesizer.isSpeaking
    }
    
    func reloadSynthesizer() {
        synthesizer = AVSpeechSynthesizer()
        synthesizer.delegate = self
    }
    
    func isPaused() -> Bool {
        return self.synthesizer.isPaused
    }
}
